openucx/ucx
# UCT/MM: Add user namespace handling for posix and cma transports

## What
Add implicit user namespace reachability checks for posix and cma transports to allow proper transport selection. Failing to detect transport unreachability leads to error, where we should instead fallback on other transport.

Issue on posix transport:
```
mm_posix.c:194  UCX  ERROR open(file_name=/proc/4382/fd/23 flags=0x0) failed: No such file or directory
```
Issue on cma transport:
```
cma_ep.c:84   process_vm_readv(pid=75643 {0x2552940,12480}-->{0x1be29e0,12480}) returned -1: Operation not permitted
```

## Why ?
The cma transport needs ptrace capability and the posix transport, when using `/proc/<pid>/fd/`, also requires that capability. In short the ptrace capability is local to the current user namespace (and it's children user namespaces).

UCX current reachability checks for relevant intra-node transports:
- sysv: IPC namespace matching
- posix: IPC namespace and PID namespace matching
- cma: IPC namespace and PID namespace matching

When two UCX instances are executed in different:
- IPC namespaces: rc can be used as a fallback for instance, if available
- PID namespaces: we detect the posix/cma unreachability and fallback properly on sysv transport.
- USER namespaces: we fail to detect unreachability, try to use cma/posix and terminate on error.

posix transport has two modes:
- use proc link (default): needs ptrace to use `/proc/<pid>/fd/`: it avoids potential lefovers
- else: standard shm, could have leftovers on some cases

The posix non-proc link mode does not need ptrace, so it can work even in different user namespaces.


## How ?
Reachability check needs to actually try to establish the connection.

Proposed handling when user namespaces are different:
- posix: don't try to fallback on non-proc-link, declare unrechability to fallback on sys
  - simpler as the info might not be available in time, sysv is very close, unrelated md alloc can still take place on posix
  - it remains possible to disable use_proc_link if posix is really needed
- cma: use dummy cma read to detect reachability to exercise capability
  - in case one of the two is a user namespace parent, reachability can return true and false
    - could lead to ep reconfiguration error
    - not really practical case

### Tests

Using `UCX_TLS=posix,cma`, both lanes must be usable:
- Manually tested starting from the same non default user namespace
- Manually tested from default user namespace

Adding user namespaces Azure tests below.

With same non-default namespaces:
- check that each transport reports reachability

With different user namespaces:
- posix transport:
  - non-proc link: use posix
  - proc link: fallback on sysv
- cma transport:
  - fallback on sysv

With different pid namespaces:
- posix/cma transport: make sure to fallback on sysv

Test output:
```
==== Running perftest namespace positive tests ====
==== Running perftest PID namespace test for posix ====
==== Running perftest PID namespace test for cma ====
==== Running perftest USER namespace test for posix ====
==== Running perftest USER namespace test for cma ====
==== Running perftest USER namespace test for posix non proc link ====
```
